internals_fsutils.go

```diff
@@ -391,53 +386,51 @@ func createZip(archiveName string, files map[string][]byte) error {
 		if err != nil {
 			return err
 		}
-		_, err = f.Write([]byte(fcont))
+		_, err = io.Copy(f, fcont)
 		if err != nil {
 			return err
 		}
 	}
 
-	// Make sure to check the error on Close.
-	err := w.Close()
-	if err != nil {
-		return err
-	}
-
-	err = ioutil.WriteFile(archiveName, buf.Bytes(), defaultFilePermissions)
-	if err != nil {
-		return err
-	}
-
-	return nil
+	// The Close call writes the checksum, so return the error
+	// instead of just deferring it.
+	return w.Close()
 }
 
-func createTar(files map[string][]byte) ([]byte, error) {
+func createTar(files map[string]io.Reader) ([]byte, error) {
 
 	// Create a buffer to write our archive to.
 	tarBuffer := new(bytes.Buffer)
 	tarWriter := tar.NewWriter(tarBuffer)
 
-	for fpath, fcont := range files {
+	for fpath, freader := range files {
+
+		fcontents, err := ioutil.ReadAll(freader)
```

Comment:
Unfortunately we need to know the size of the buffer we're about to write in order to write the header below, so naive solution is to read the bytes back onto the heap, but at least this specific function now keeps those bytes only for each file in the for loop, so the GC can clean up if it needs to after each iteration.